libs/geometry/include/geometry/SurfaceOrientation.h

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (C) 2019 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef TNT_GEOMETRY_SURFACEORIENTATION_H
+#define TNT_GEOMETRY_SURFACEORIENTATION_H
+
+#include <math/quat.h>
+#include <math/vec3.h>
+#include <math/vec4.h>
+
+namespace filament {
+namespace geometry {
+
+struct OrientationBuilderImpl;
+struct OrientationImpl;
+
+/**
+ * The surface orientation helper can be used to populate Filament-style TANGENTS buffers.
+ */
+class SurfaceOrientation {
+public:
+
+    /**
+     * Constructs an immutable surface orientation helper.
+     *
+     * At a minimum, clients must supply a vertex count and normals buffer. They can supply data in
+     * any of the following three combinations:
+     *
+     *   1. vec3 normals only (not recommended)
+     *   2. vec3 normals + vec4 tangents (sign of W determines bitangent orientation)
+     *   3. vec3 normals + vec2 uvs + vec3 positions + uint3 indices
+     */
+    class Builder {
+    public:
+        Builder();
+        ~Builder() noexcept;
+
+        /**
+         * These two attributes are required. They are not passed into the constructor to force
+         * calling code to be self-documenting.
+         * @{
+         */
+        Builder& vertexCount(size_t vertexCount) noexcept;
+        Builder& normals(const filament::math::float3*) noexcept;
+        /** @} */
+
+        Builder& tangents(const filament::math::float4*) noexcept;
+        Builder& uvs(const filament::math::float2*) noexcept;
+        Builder& positions(const filament::math::float3*) noexcept;
+
+        Builder& triangleCount(size_t triangleCount) noexcept;
+        Builder& triangles(const filament::math::uint3*) noexcept;
+        Builder& triangles(const filament::math::ushort3*) noexcept;
+
+        /**
+         * Supplying explicit stride values is optional, they each default to the size of their
+         * respective input attributes, e.g. sizeof(float3) for normals.
+         * @{
+         */
+        Builder& normalStride(size_t numBytes) noexcept;
+        Builder& tangentStride(size_t numBytes) noexcept;
+        Builder& uvStride(size_t numBytes) noexcept;
+        Builder& positionStride(size_t numBytes) noexcept;
+        /** @} */
+
+        /**
+         * Generates quats or panics if the submitted data is an incomplete combination.
+         */
+        SurfaceOrientation* build();
```

Comment:
This should return `SurfaceOrientation` by value, this way we wouldn't need the `destroy()` static. This would work because `SurfaceOrientation` is really just a pointer.  You'd have to allow the move-ctor and move operator.
This would make sure the caller doesn't forget to call destroy().

Reply:
Done.